src/md5/md5.cc

```diff
@@ -250,11 +235,9 @@ MD5::MD5(ifstream& stream){
 
 unsigned char *MD5::raw_digest(){
 
-	uint8_t* s = new uint8_t[16];
+  unsigned char *s = new unsigned char[16];
 
   if (!finalized){
-    cerr << "MD5::raw_digest:  Can't get digest if you haven't "<<
-      "finalized the digest!" <<endl;
```

Comment:
Aren't those error messages a good thing? Now the caller doesn't have an idea why it failed.

Reply:
This is sort of a library and it shouldn't produce unexpected console output. Plus this doesn't work with a windows GUI application anyway. Plus this doesn't help you to catch this error programmatically.

On second thought, this isn't even a real error condition but a programmer error that does only deserve an assertion, like some code below already does. It's a much more serious error to actually ignore the read errors from the FILE/stream.

It's just amazing how many issues are in this tiny piece of code, i just wanted to fix some warnings...

Reply:
I see. What do you think about throwing a `std::runtime_error`?

Reply:
The exception for what? Getting called in the wrong finalization state? Or the read errors? I wouldn't use an exception here for both cases, this whole "library" is exception free and looks so much c-like and the calling side is not prepared for this, its not worth the trouble. Assert for the finalization and if necessary use a return code for read errors.

Reply:
The calling side isn't prepared for an assert either though. And what "trouble" do you mean?

Reply:
Maybe there is a misunderstanding what an assert is, so lets clear this up: In release mode an assert does nothing, in debug mode, if it triggers, it displays the message and then terminates the program, so there is no calling side anymore ;).

wxWidgets has these nice "hybrid-asserts", they are also active in release mode and silently perform the trigger-action there, in debug mode they pop up the message and let you choose what to do. These are a "nicer" version of a "true" assert, they allow the program to continue with some meaningful value in cause of the assert.

The "trouble" with exceptions here is that they are not so common like in java and here there is also no such thing like checked exceptions. So you have to write some documentation like "This does throw this and that exception in case of error". The TiCPP library does this, wxFB itself uses internally its own exception class, but this class is so small, does nasty c-style things like returning something you need to delete, basically the only c++-element is the istream (ok, and ostream inserter), an error code would fit here much better in my opinion.

With some small modifications you don't even need to signal read errors at all, you could let the caller check the error state of the stream to detect an error (currently it always will be in error state because of the way it is read from, but this can be made better). For the FILE version you need to stop to delete it in the method and it would work there as well. That would give a nice interface that is identical to the memory block version.